src/workerd/util/sqlite-test.c++

```diff
@@ -1076,5 +1076,199 @@ KJ_TEST("SQLite onRollback") {
+    SELECT id, value FROM things;
+  )"));
+
+  {
+    auto query = stmt.run();
+
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 1);
+    KJ_ASSERT(query.getInt(1) == 123);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 2);
+    KJ_ASSERT(query.getInt(1) == 456);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 3);
+    KJ_ASSERT(query.getInt(1) == 789);
+    query.nextRow();
+
+    KJ_ASSERT(query.isDone());
+  }
+
+  {
+    auto query = stmt.run();
+
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 1);
+    KJ_ASSERT(query.getInt(1) == 123);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 2);
+    KJ_ASSERT(query.getInt(1) == 456);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 3);
+    KJ_ASSERT(query.getInt(1) == 789);
+    query.nextRow();
+
+    // Re-running the statement inserted duplicates, so we'll see those in the results.
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 4);
+    KJ_ASSERT(query.getInt(1) == 123);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 5);
+    KJ_ASSERT(query.getInt(1) == 456);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 6);
+    KJ_ASSERT(query.getInt(1) == 789);
+    query.nextRow();
+
+    KJ_ASSERT(query.isDone());
+  }
+
+  // Test resetting the database, which will force re-parsing each statement.
+  db.reset();
+
+  {
+    auto query = stmt.run();
+
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 1);
+    KJ_ASSERT(query.getInt(1) == 123);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 2);
+    KJ_ASSERT(query.getInt(1) == 456);
+    query.nextRow();
+    KJ_ASSERT(!query.isDone());
+    KJ_ASSERT(query.getInt(0) == 3);
+    KJ_ASSERT(query.getInt(1) == 789);
+    query.nextRow();
+
+    KJ_ASSERT(query.isDone());
+  }
+}
+
+KJ_TEST("SQLite prepareMulti with failure") {
+  // Test running a multi-line prepared statement that fails in the middle.
+
+  auto dir = kj::newInMemoryDirectory(kj::nullClock());
+  SqliteDatabase::Vfs vfs(*dir);
+  SqliteDatabase db(vfs, kj::Path({"foo"}), kj::WriteMode::CREATE | kj::WriteMode::MODIFY);
+
+  auto stmt = db.prepareMulti(SqliteDatabase::TRUSTED, kj::str(R"(
+    CREATE TABLE IF NOT EXISTS things (
+      id INTEGER PRIMARY KEY AUTOINCREMENT,
+      value INTEGER
+    );
+    INSERT INTO things(value) VALUES (123);
+    INSERT INTO things(id, value) VALUES (1, 456);  -- fails, duplicate primary key
+  )"));
+
+  KJ_EXPECT_THROW_MESSAGE("SQLITE_CONSTRAINT", stmt.run());
+  KJ_EXPECT_THROW_MESSAGE("SQLITE_CONSTRAINT", stmt.run());
+  KJ_EXPECT_THROW_MESSAGE("SQLITE_CONSTRAINT", stmt.run());
+
+  // We ran the statement three times. Each time it should have inserted a new row containing
```

Comment:
We talked offline yesterday about `exec()`ing multiple statements and how if one fails in the middle, we should probably roll back all of the statements.  For my own education, in a world where we fix that bug, would `prepareMulti` implement the rollback or would something higher than `prepareMulti` handle the rollback?

If it's prepareMulti, should we drop a TODO(soon) here about how even this behavior is buggy?

Reply:
Technically `prepareMulti()` itself doesn't parse or execute anything. But I think your question is, would `SqliteDatabase` handle the rollback internally or would the caller have to do it. I think `SqliteDatabase` should handle it. I suspect logic will be needed in all of `prepareSql()`, `Statement::prepareForExecution()`, and `~Query()`, annoyingly.

> If it's prepareMulti, should we drop a TODO(soon) here about how even this behavior is buggy?

Added comment:

https://github.com/cloudflare/workerd/pull/2970/files#diff-2b573b1c6fa7749e40c7da2c29f37f0f4accf53f2406bdab6ca5378755260fecR1173-R1177

(No other changes, just rebased.)